### PR TITLE
Feature/maroon 303 throwing balls experiment

### LINKED ISSUE #303: [3D Motion Simulation] Implementation of 3D Motion Calculation Sezanrios
3D Motion Calculation for:

- Default Setting
- Satellite
- Ball in the Wind
- Charged Particle
- Resonance
- Parametric Oscillator
- Relaxation Oscillator

### PR DESCRIPTION
**Pull request by lebsar**
closes #303

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -124,3 +124,5 @@ ExportedObj/
 
 # IntelliJ
 .idea/
+
+unity/ProjectSettings/ProjectSettings.asset
```
Comment (michaelholly) [inline .gitignore]: Please do not add the project settings into the gitignore file.

File: unity/Assets/3rd Party/ncalc/LICENSE [hunk 2]
```diff
@@ -0,0 +1,21 @@
+MIT License
```
Comment (michaelholly) [inline unity/Assets/3rd Party/ncalc/LICENSE]: Please move the 3rd party stuff in the new vendor folder structure.

File: unity/Assets/Ball.cs [hunk 3]
```diff
@@ -0,0 +1,18 @@
+﻿using System.Collections;
```
Comment (michaelholly) [inline unity/Assets/Ball.cs]: Why do we need this empty file?
If it is useless please delete it.

File: unity/Assets/CoordinateSystemInit.cs [hunk 4]
```diff
@@ -0,0 +1,55 @@
+﻿using UnityEngine;
```
Comment (michaelholly) [inline unity/Assets/CoordinateSystemInit.cs]: Please move this file into your experiment folder.

File: .gitignore (hunk 1 above)
Comment (codecreativity) [inline .gitignore]: That's probably a bad idea

**Review by codecreativity — DISMISSED** [review]

File: unity/Assets/Maroon/reusablePrefabs/Assets MOOD WARE/_MW Skybox/MAT/Skybox 22_pan.mat [hunk 5]
```diff
@@ -0,0 +1,61 @@
+%YAML 1.1
```
Comment (michaelholly) [inline unity/Assets/Maroon/reusablePrefabs/Assets MOOD WARE/_MW Skybox/MAT/Skybox 22_pan.mat]: Please move the skybox stuff to your experiment folder e.g. Materials.
Delete all unused or unnecessary files from the asset folder.
Rename the asset folder to something meaningful.

File: unity/Assets/Maroon/reusablePrefabs/DeepSpaceSkyboxPack/DiverseSpace/Material/DiverseSpaceMaterial.mat [hunk 6]
```diff
@@ -0,0 +1,53 @@
+%YAML 1.1
```
Comment (michaelholly) [inline unity/Assets/Maroon/reusablePrefabs/DeepSpaceSkyboxPack/DiverseSpace/Material/DiverseSpaceMaterial.mat]: Please move this skybox stuff into your experiment folder e.g. Materials.
Delete all unused or unnecessary files from the asset folder.

File: unity/Assets/Maroon/reusablePrefabs/MilkyWay/Material/MilkyWay.mat [hunk 7]
```diff
@@ -0,0 +1,113 @@
+%YAML 1.1
```
Comment (michaelholly) [inline unity/Assets/Maroon/reusablePrefabs/MilkyWay/Material/MilkyWay.mat]: Do we still need this skybox?
If not, please delete it.

File: unity/Assets/Maroon/reusablePrefabs/Planet Earth Free/Demo.unity [hunk 8]
```diff
@@ -0,0 +1,348 @@
+%YAML 1.1
```
Comment (michaelholly) [inline unity/Assets/Maroon/reusablePrefabs/Planet Earth Free/Demo.unity]: Do we need this stuff?
Please delete it.

File: unity/Assets/Maroon/reusablePrefabs/Planet Earth Free/Materials/SkyboxMaterial.mat [hunk 9]
```diff
@@ -0,0 +1,104 @@
+%YAML 1.1
```
Comment (michaelholly) [inline unity/Assets/Maroon/reusablePrefabs/Planet Earth Free/Materials/SkyboxMaterial.mat]: unused?

File: unity/Assets/Maroon/reusablePrefabs/Soccer Ball/Prefabs/Soccer Ball.prefab [hunk 10]
```diff
@@ -0,0 +1,96 @@
+%YAML 1.1
```
Comment (michaelholly) [inline unity/Assets/Maroon/reusablePrefabs/Soccer Ball/Prefabs/Soccer Ball.prefab]: Please move the soccer ball prefab to your experiment prefab folder.

File: unity/Assets/Maroon/reusablePrefabs/magellan/Magellan.unity [hunk 11]
```diff
@@ -0,0 +1,366 @@
+%YAML 1.1
```
Comment (michaelholly) [inline unity/Assets/Maroon/reusablePrefabs/magellan/Magellan.unity]: I think we do not need this scene file.
Please delete it.

File: unity/Assets/Maroon/reusablePrefabs/magellan/maggellan_ex.prefab [hunk 12]
```diff
@@ -0,0 +1,658 @@
+%YAML 1.1
```
Comment (michaelholly) [inline unity/Assets/Maroon/reusablePrefabs/magellan/maggellan_ex.prefab]: Please move it to your experiment prefab folder and rename it to satellite

File: unity/ProjectSettings/ProjectSettings.asset [hunk 13]
```diff
@@ -125,7 +125,8 @@ PlayerSettings:
     16:9: 1
     Others: 1
   bundleVersion: 20210709
-  preloadedAssets: []
+  preloadedAssets:
```
Comment (michaelholly) [inline unity/ProjectSettings/ProjectSettings.asset]: Please revert these changes!

File: unity/ProjectSettings/EditorBuildSettings.asset [hunk 14]
```diff
@@ -5,73 +5,76 @@ EditorBuildSettings:
   m_ObjectHideFlags: 0
```
Comment (michaelholly) [inline unity/ProjectSettings/EditorBuildSettings.asset]: Please do not deactivate all other scenes in the build settings!

File: unity/Assets/skybox_earth.jpg.meta [hunk 15]
```diff
@@ -0,0 +1,85 @@
+fileFormatVersion: 2
```
Comment (michaelholly) [inline unity/Assets/skybox_earth.jpg.meta]: Please delete it if it is unused!

File: unity/Assets/StreamingAssets/Assessment/ThrowingBalls.meta [hunk 16]
```diff
@@ -0,0 +1,8 @@
+fileFormatVersion: 2
```
Comment (michaelholly) [inline unity/Assets/StreamingAssets/Assessment/ThrowingBalls.meta]: This stuff should not be part of the assessment system. Please move it to another folder in the streaming assets.
e.g. /StreamingAssets/ExperimentParameters/3DMotionSimulation/

File: unity/Assets/Maroon/scenes/templates/ExperimentRoomTemplate.pc/LightingData.asset.meta [hunk 17]
```diff
@@ -1,5 +1,5 @@
 fileFormatVersion: 2
-guid: 4a3d56e541f04a54d82db9bdbaabbec3
```
Comment (michaelholly) [inline unity/Assets/Maroon/scenes/templates/ExperimentRoomTemplate.pc/LightingData.asset.meta]: Please revert the ExperimentRoomTemplate changes.

File: unity/Assets/Maroon/scenes/templates/ExperimentRoomTemplate.pc/Lightmap-0_comp_dir.png.meta [hunk 18]
```diff
@@ -1,5 +1,5 @@
 fileFormatVersion: 2
-guid: a451ad8f14848f642b656337f569057e
```
Comment (michaelholly) [inline unity/Assets/Maroon/scenes/templates/ExperimentRoomTemplate.pc/Lightmap-0_comp_dir.png.meta]: same here. revert it

File: unity/Assets/Maroon/scenes/templates/ExperimentRoomTemplate.pc/Lightmap-0_comp_light.exr.meta [hunk 19]
```diff
@@ -1,5 +1,5 @@
 fileFormatVersion: 2
-guid: c995292da3ade0d4f860fcc57de8ed9d
```
Comment (michaelholly) [inline unity/Assets/Maroon/scenes/templates/ExperimentRoomTemplate.pc/Lightmap-0_comp_light.exr.meta]: same here. revert it

File: unity/Assets/Maroon/scenes/templates/ExperimentRoomTemplate.pc/Lightmap-0_comp_shadowmask.png.meta [hunk 20]
```diff
@@ -1,5 +1,5 @@
 fileFormatVersion: 2
-guid: 2a7f43f40d0434f438e84dcd258855b4
```
Comment (michaelholly) [inline unity/Assets/Maroon/scenes/templates/ExperimentRoomTemplate.pc/Lightmap-0_comp_shadowmask.png.meta]: same here. revert it

File: unity/Assets/Maroon/scenes/templates/ExperimentRoomTemplate.pc/ReflectionProbe-0.exr.meta [hunk 21]
```diff
@@ -1,5 +1,5 @@
 fileFormatVersion: 2
-guid: b15e4cb00d7ffd144a1d5ef21a4d92e3
```
Comment (michaelholly) [inline unity/Assets/Maroon/scenes/templates/ExperimentRoomTemplate.pc/ReflectionProbe-0.exr.meta]: same here. revert it

File: unity/Assets/Maroon/scenes/templates/Resources/ExperimentRoomTemplate.laboratoryblock.prefab.meta [hunk 22]
```diff
@@ -1,5 +1,5 @@
 fileFormatVersion: 2
-guid: 43db3f423f16b2c4b8829cd6c4677ff0
```
Comment (michaelholly) [inline unity/Assets/Maroon/scenes/templates/Resources/ExperimentRoomTemplate.laboratoryblock.prefab.meta]: same here. revert it

File: unity/Assets/Maroon/scenes/experiments/ThrowingBalls/Resources/ExperimentRoomTemplate.laboratoryblock.prefab [hunk 23]
```diff
@@ -0,0 +1,71 @@
+%YAML 1.1
```
Comment (michaelholly) [inline unity/Assets/Maroon/scenes/experiments/ThrowingBalls/Resources/ExperimentRoomTemplate.laboratoryblock.prefab]: Please rename this file according to your experiment name.

File: unity/Assets/Maroon/scenes/experiments/ThrowingBalls/Resources/BallInTheWind.json [hunk 24]
```diff
@@ -0,0 +1,19 @@
+[
+	{
```
Comment (michaelholly) [inline unity/Assets/Maroon/scenes/experiments/ThrowingBalls/Resources/BallInTheWind.json]: Do we need the json files in the experiment resource folder when it is included in the streaming assets.

File: unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts.meta [hunk 25]
```diff
@@ -0,0 +1,8 @@
+fileFormatVersion: 2
```
Comment (michaelholly) [inline unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts.meta]: can we rename the folder to "Scripts"?

File: unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/Calculation.cs [hunk 26]
```diff
@@ -0,0 +1,611 @@
+﻿using System.Collections;
```
Comment (michaelholly) [inline unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/Calculation.cs]: remove unused directives (System.Collections and Maroon.UI)

File: unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/Calculation.cs [hunk 27]
```diff
@@ -0,0 +1,611 @@
...
+using System.Collections.Generic;
+using UnityEngine;
+using NCalc;
+using Maroon.UI;
+using ObjectsInUse;
+
+public class Calculation : PausableObject, IResetObject
```
Comment (michaelholly) [inline unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/Calculation.cs]: we should rename this class and move it to a namespace
e.g. namespace Maroon.Physics.3DMotion

File: unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/Calculation.cs [hunk 28]
```diff
@@ -0,0 +1,611 @@
...
+    private Vector3 _point;
+    
+    private bool _stopSimulation = false;
+    private bool _startCalcPlot = false;
+    private int _currentSteps = 0;
+    private int _currentUpdateRate = 0;
+    private bool _drawTrajectory = true;
+    private bool _firstIteration = true;
+    private bool _initialized = false;
+    private const float _threshold = 0.001f;
```
Comment (michaelholly) [inline unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/Calculation.cs]: Const values => capitalized
private const float Threshold = 0.001f;

File: unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/Calculation.cs [hunk 29]
```diff
@@ -0,0 +1,611 @@
...
+    private List<Vector2> _dataFz = new List<Vector2>();
+
+    private List<Vector2> _dataP = new List<Vector2>();
+    private List<Vector2> _dataEkin = new List<Vector2>();
+    private List<Vector2> _data_W = new List<Vector2>();
+
+    /// <summary>
+    /// Initialization
+    /// </summary>
+    protected override void Start()
```
Comment (michaelholly) [inline unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/Calculation.cs]: The unused method which overrides the Start method of PausableObject?
Can be removed?

File: unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/Calculation.cs [hunk 30]
```diff
@@ -0,0 +1,611 @@
...
+        _currentFx = 0;
+        _currentFy = 0;
+        _currentFz = 0;
+
+        _currentP = 0;
+        _currentEkin = 0;
+        _currentW = 0;
+        _oldPower = 0;
+
+        _xMax = System.Int64.MinValue;
```
Comment (michaelholly) [inline unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/Calculation.cs]: Can be simplified.
e.g. long.MinValue

File: unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/Calculation.cs [hunk 31]
```diff
@@ -0,0 +1,611 @@
...
+
+    private float _xMin = System.Int64.MaxValue;
+    private float _yMin = System.Int64.MaxValue;
+    private float _zMin = System.Int64.MaxValue;
+
+    private string _formulaFx; 
+    private string _formulaFy; 
+    private string _formulaFz; 
+
+    private double _currentFx = 0;
```
Comment (michaelholly) [inline unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/Calculation.cs]: do we need double values for the calculation?

File: unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/Calculation.cs [hunk 32]
```diff
@@ -0,0 +1,611 @@
...
+    /// Evaluates the formula and handles errors while evaluating
+    /// </summary>
+    private void CalcForces()
+    {
+        Expression e;
+
+        try
+        {
+            // calc x-force
+            if (!System.String.IsNullOrEmpty(_formulaFx))
```
Comment (michaelholly) [inline unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/Calculation.cs]: can be simplified with built-in type string

File: unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/Calculation.cs [hunk 33]
```diff
@@ -0,0 +1,611 @@
...
+    /// <summary>
+    /// Inits the parameters from the GUI 
+    /// </summary>
+    private void InitParameter()
+    {
+        _formulaFx = ParameterUI.Instance.GetFunctionFx();
+        _formulaFy = ParameterUI.Instance.GetFunctionFy();
+        _formulaFz = ParameterUI.Instance.GetFunctionFz();
+        
+        _mass = (double) ParameterUI.Instance.GetMass();
```
Comment (michaelholly) [inline unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/Calculation.cs]: typecast is redundant.

File: unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/Calculation.cs [hunk 34]
```diff
@@ -0,0 +1,611 @@
...
+                _yMax = (float)_currentY;
+            }
+            if (_zMax < _currentZ)
+            {
+                _zMax = (float)_currentZ;
+            }
+        }
+        
+        // add scaling for nice visualization
+        _xMin = _xMin - GetMinMaxScaleFactor(_xMin);
```
Comment (michaelholly) [inline unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/Calculation.cs]: Use compound assignment
"-=" or "+="

File: unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/Calculation.cs [hunk 35]
```diff
@@ -0,0 +1,611 @@
...
+    /// <summary>
+    /// This method checks if scaling is needed
+    /// If a certain value is smaller than the threshold -> scaling leads to 
+    /// incorrect visualization
+    /// </summary>
+    /// <param name="value">Value to check</param>
+    /// <returns>Bool if scaling is needed</returns>
+    private bool ScalingNeeded(float value)
+    {
+        if (System.Math.Abs(value) < _threshold)
```
Comment (michaelholly) [inline unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/Calculation.cs]: Convert it to return statement
e.g.  return !(System.Math.Abs(value) < Threshold);

File: unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/Calculation.cs [hunk 36]
```diff
@@ -0,0 +1,611 @@
...
+    /// <summary>
+    /// Method to scale min/max values for visualization
+    /// </summary>
+    /// <param name="value">Value to scale</param>
+    /// <returns>Adapted value</returns>
+    private float GetMinMaxScaleFactor(float value)
+    {
+        value = System.Math.Abs(value);
+
+        if (value == 0)
```
Comment (michaelholly) [inline unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/Calculation.cs]: Convert it to return statement
return value == 0 ? 0.5f : 0f;

File: unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/CoordCameraController.cs [hunk 37]
```diff
@@ -0,0 +1,103 @@
+using System;
```
Comment (michaelholly) [inline unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/CoordCameraController.cs]: Remove unused directives

File: unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/CoordSystem.cs [hunk 38]
```diff
@@ -0,0 +1,385 @@
...
+using System.Collections.Generic;
+using UnityEngine;
+using TMPro;
+using ObjectsInUse;
+
+public class CoordSystem : MonoBehaviour, IResetObject
+{
+    private static CoordSystem _instance;
+    
+    [SerializeField] private Transform _coordOrigin;
```
Comment (michaelholly) [inline unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/CoordSystem.cs]: _coordOrigin is assigned to the transform object itself in the inspector view.
why are you not using e.g. transform.position or transform.rotation in the code?

File: unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/CoordSystem.cs [hunk 39]
```diff
@@ -0,0 +1,385 @@
...
+    [SerializeField] private GameObject _trajectory;
```
Comment (michaelholly) [inline unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/CoordSystem.cs]: for assigning in the inspector use a prefab instead of a game object in the scene.

File: unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/CoordSystem.cs [hunk 40]
```diff
@@ -0,0 +1,385 @@
...
+    [SerializeField] private GameObject _particle;
```
Comment (michaelholly) [inline unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/CoordSystem.cs]: do we need 4 variables for this?
why are you not using one game object for the particle object? this object can be set regarding the settings.
this allows us to extend it more easily.

File: unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/CoordSystem.cs [hunk 41]
```diff
@@ -0,0 +1,385 @@
+﻿using System.Collections;
```
Comment (michaelholly) [inline unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/CoordSystem.cs]: remove unused directives.

File: unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/CoordSystem.cs [hunk 42]
```diff
@@ -0,0 +1,385 @@
...
+        _trajectory.transform.localScale = scaleTrajectory;
+        _particle.transform.localScale = scaleParticle;
+        _satellite.transform.localScale = scaleParticle;
+        _planet.transform.localScale = scalePlanet;
+
+        DrawAxis();
+    }
+
+    // Update is called once per frame
+    void Update()
```
Comment (michaelholly) [inline unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/CoordSystem.cs]: For performance reasons, delete empty update function calls.

File: unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/CoordSystem.cs [hunk 43]
```diff
@@ -0,0 +1,385 @@
...
+    {
+    }
+
+    /// <summary>
+    /// Inits the max/min range of the coordinate system
+    /// </summary>
+    void DrawAxis()
+    {
+        Vector3 end;
+        float origin_x = _coordOrigin.position.x;
```
Comment (michaelholly) [inline unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/CoordSystem.cs]: just transform.position.x?

File: unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/CoordSystem.cs [hunk 44]
```diff
@@ -0,0 +1,385 @@
...
+    /// <param name="start">Start points of the lines</param>
+    /// <param name="end">End points of the lines</param>
+    /// <param name="color">Colour of the lines</param>
+    void DrawOrigin(Vector3 start, Vector3 end, Color color)
+    {
+        GameObject myLine = new GameObject();
+        myLine.transform.position = start;
+        myLine.AddComponent<LineRenderer>();
+        LineRenderer lr = myLine.GetComponent<LineRenderer>();
+        lr.material = new Material(Shader.Find("Specular"));
```
Comment (michaelholly) [inline unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/CoordSystem.cs]: for performance reasons, do not create a material with new!
create a new material and assign it in the inspector.

File: unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/CoordSystem.cs [hunk 45]
```diff
@@ -0,0 +1,385 @@
...
+    }
+
+    /// <summary>
+    /// Rounds calculated values into the form x.xx for visualization
+    /// </summary>
+    /// <param name="value">Original value</param>
+    /// <returns>Rounded value</returns>
+    private string RoundDisplayedValues(float value)
+    {
+        if (System.Math.Round(value, 2) == 0.00f)
```
Comment (michaelholly) [inline unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/CoordSystem.cs]: convert it to ?:

File: unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/FileController.cs [hunk 46]
```diff
@@ -0,0 +1,126 @@
+﻿using System.Collections;
```
Comment (michaelholly) [inline unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/FileController.cs]: delete unused directives

File: unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/FileController.cs [hunk 47]
```diff
@@ -0,0 +1,126 @@
...
+using System.Collections.Generic;
+using UnityEngine;
+using Newtonsoft.Json;
+using System.IO;
+using Maroon.GlobalEntities;
+
+public class FileController : MonoBehaviour
```
Comment (michaelholly) [inline unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/FileController.cs]: maybe find another name for the class
e.g ParameterLoader

File: unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/FileController.cs [hunk 48]
```diff
@@ -0,0 +1,126 @@
...
+        public float Y;
+        public float Z;
+        public float Vx;
+        public float Vy;
+        public float Vz;
+
+    }
+
+    // Start is called before the first frame update
+    void Start()
```
Comment (michaelholly) [inline unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/FileController.cs]: use the explicit private modifier

File: unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/FileController.cs [hunk 49]
```diff
@@ -0,0 +1,126 @@
...
+    {
+        // Listener for extern json data 
+        WebGlReceiver.Instance.OnIncomingData.AddListener(data =>
+        {
+            LoadExternJsonFile(data);
```
Comment (michaelholly) [inline unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/FileController.cs]: replace it with method group
e.g. WebGlReceiver.Instance.OnIncomingData.AddListener(LoadExternJsonFile);

File: unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/MaxCameraPosition.cs [hunk 50]
```diff
@@ -0,0 +1,38 @@
...
+
+    // Start is called before the first frame update
+    void Start()
+    {
+        _startPosition = transform.position;
+    }
+
+    public void ChangePosition()
+    {
+        transform.position = new Vector3(9.7f, 7.1f, 12.7f);
```
Comment (michaelholly) [inline unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/MaxCameraPosition.cs]: is it a good idea to assign this position hardcoded?
and maybe use the CoordCameraController for changing the position for max and min
I think we can remove the MaxCameraPosition and MinCameraPosition script in this case

File: unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/MinCameraPosition.cs [hunk 51]
```diff
@@ -0,0 +1,38 @@
+﻿using System.Collections;
```
Comment (michaelholly) [inline unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/MinCameraPosition.cs]: same as for the MaxCameraPosition

File: unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/ParameterUI.cs [hunk 52]
```diff
@@ -0,0 +1,621 @@
...
+{
+    public enum ParticleObject
+    {
+        Default,
+        Ball,
+        Satellite
+    }
+}
+
+public class ParameterUI : PausableObject, IResetObject
```
Comment (michaelholly) [inline unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/ParameterUI.cs]: for maintenance, this class is a very bad idea
e.g. don't use UI references in the code. use e.g. the onvalue change event to handle the input
please contact me in this case so that we can resolve this issue.

File: unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/SkyboxController.cs [hunk 53]
```diff
@@ -0,0 +1,90 @@
+﻿using System.Collections;
```
Comment (michaelholly) [inline unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/SkyboxController.cs]: remove unused directives

File: unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/SkyboxController.cs [hunk 54]
```diff
@@ -0,0 +1,90 @@
...
+public class SkyboxController : MonoBehaviour, IResetObject
+{
+    private static SkyboxController _instance;
+
+    [SerializeField] private Material _grass;
+    [SerializeField] private Material _space;
+    [SerializeField] private Material _experimentRoom;
+
+    // Start is called before the first frame update
+    void Start()
```
Comment (michaelholly) [inline unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/SkyboxController.cs]: remove empty Start and Update

File: unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/SkyboxController.cs [hunk 55]
```diff
@@ -0,0 +1,90 @@
...
+        
+    }
+
+    /// <summary>
+    /// Sets the background of the scene
+    /// </summary>
+    /// <param name="background">Background to use</param>
+    public void SetBackground(string background)
+    {
+        if (background == "ExperimentRoom")
```
Comment (michaelholly) [inline unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/SkyboxController.cs]: Convert it to switch

File: unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/ValueGraph.cs [hunk 56]
```diff
@@ -0,0 +1,180 @@
+﻿using System.Collections;
```
Comment (michaelholly) [inline unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/ValueGraph.cs]: delete unused directives

File: unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/ValueGraph.cs [hunk 57]
```diff
@@ -0,0 +1,180 @@
...
+
+    /// <summary>
+    /// Loads and displays the choosen value in the LineChart (GUI)
+    /// </summary>
+    /// <param name="choice">Data to load</param>
+    public void LoadChoosenValue(int choice)
+    {
+        switch (choice)
+        {
+            case 0:
```
Comment (michaelholly) [inline unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/ValueGraph.cs]: use a enum for the cases

File: unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/ValueGraph.cs [hunk 58]
```diff
@@ -0,0 +1,180 @@
...
+public class ValueGraph : PausableObject, IResetObject
+{
+
+    private static ValueGraph _instance;
+    public SimpleLineChart simpleLineChart;
+
+    /// <summary>
+    /// Initialization
+    /// </summary>
+    protected override void Start()
```
Comment (michaelholly) [inline unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/ValueGraph.cs]: remove empty Start method

**Review by michaelholly — CHANGES_REQUESTED** [review]
Many thanks for this changes.
For merging this PR we request a few changes. Please see the comments.

Please rename also the experiment to e.g. 3DMotionSimulation or 3DMotionExperiment
Furthermore, the experiment laboratory block prefab is missing
In the end, all experiment relevant objects should be included in one prefab so that the experiment can be used in other scenes (e.g. VR) too.

File: unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/ParameterUI.cs (hunk 52 above)
Comment (lebsar) [reply unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/ParameterUI.cs]: I would like to discuss this with you in our next meeting.

Comment (lebsar) [reply unity/Assets/Maroon/scenes/experiments/ThrowingBalls/scripts/ParameterUI.cs]: As we discussed today (10.11.21): We will leave the ParameterUI class as it is for the moment and create a future-issue to refactor this class in the future.

**Review by michaelholly — APPROVED** [review]
